frame/pools/src/lib.rs

```diff
@@ -0,0 +1,1135 @@
+//! * Delegators cannot quickly transfer to another pool if they do no like nominations, instead
+//!   they must wait for the unbonding duration.
+//!
+//! # Pool creation and upkeep
+//!
+//! TBD - possible options:
+//! * Pools can be created by anyone but nominations can never be updated
+//! * Pools can be created by anyone and the creator can update the targets
+//! * Pools are created by governance and governance can update the targets
+//! ... Other ideas
+//! * pools can have different roles assigned: creator (puts deposit down, cannot remove deposit
+//!   until pool is empty), admin (can control who is nominator, destroyer and can do
+//!   nominate/destroy), nominator (can adjust targets), destroyer (can initiate destroy), etc
+//!
+//! # Runtime builder warnings
+//!
+//! * watch out for overflow of [`RewardPoints`] and [`BalanceOf`] types. Consider things like the
+//!   chains total issuance, staking reward rate, and burn rate.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// TODO:
+// - make withdraw unbonded permissions-less and remove withdraw unbonded call from unbond
+// - make claiming rewards permissionless
+// - creation
+//    - CreateOrigin: the type of origin that can create a pool - can be set with governance call
+//    - creator: account that cannont unbond until there are no other pool members (essentially
+//      deposit)
+//    - kicker: can kick (force unbond) delegators and block new delegators
+//    - nominator: can set targets
+//    - admin: can change kicker, nominator, and make another account admin.
+// - checks for number of pools when creating pools (param for max pools, pool creation origin)
+// - post checks that rewardpool::count == bondedpool::count. delegators >= bondedpool::count,
+//   subpools::count <= bondedpools
+
+use frame_support::{
+	ensure,
+	pallet_prelude::*,
+	storage::bounded_btree_map::BoundedBTreeMap,
+	traits::{Currency, ExistenceRequirement, Get},
+	DefaultNoBound, RuntimeDebugNoBound,
+};
+use scale_info::TypeInfo;
+use sp_core::U256;
+use sp_runtime::traits::{Bounded, Convert, Saturating, StaticLookup, TrailingZeroInput, Zero};
+use sp_staking::{EraIndex, PoolsInterface, SlashPoolArgs, SlashPoolOut, StakingInterface};
+use sp_std::{collections::btree_map::BTreeMap, ops::Div};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub use pallet::*;
+pub(crate) const LOG_TARGET: &'static str = "runtime::pools";
+
+// Syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
+			target: LOG_TARGET,
+			concat!("[{:?}] 💰", $patter), <frame_system::Pallet<T>>::block_number() $(, $values)*
+		)
+	};
+}
+
+type PoolId = u32;
+type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+type SubPoolsWithEra<T> = BoundedBTreeMap<EraIndex, UnbondPool<T>, MaxUnbonding<T>>;
+// NOTE: this assumes the balance type u128 or smaller.
```

Comment:
could have an integrity test or `const_assert`, with explanation of why it assumes this, and why U256 is good enough.